Adding the methods polredbest/polredabs to a numberfield

edgarcosta opened the pull request:
<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->

The PR adds polynomial reduction methods to number fields. Prior to PR #39920, `polredbest` was applied internally by default, but this approach proved suboptimal for large polynomials. This change gives users the ability to explicitly call `polredbest`/`polredabs` on number fields after creation.

`optimized_representation` now also accepts `algorithm='polredbest'` (default) or `algorithm='polredabs'`, so existing callers can opt into the canonical reduction without switching to the new methods.

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

edgarcosta committed:
adding polred* methods

edgarcosta committed:
specify monic

user202729 commented:
implementation is temporary, API is forever. We may want to think more carefully about what the method is named.

`polredbest` sounds very undescriptive to me. And the external guarantee is way too strong (what if later some better library come along, are we forbidden to switch to use that library instead?)

edgarcosta commented:
I realized that `optimized_representation` is already calling `polredbest`, so we could scrape the method `polredbest`, and perhaps clarify in the documentation that we are calling `polredbest` and we are trying to minimize discriminant.

For `polredabs`, it is unclear, we could call it minimal/canonical in some sense, but the sense it is minimal/canonical, is in the sense that that `PARI` defines it:

> The precise definition of the output of [polredabs](https://pari.math.u-bordeaux.fr/dochtml/html-stable/General_number_fields.html#se:polredabs) is as follows.
> 
> * Consider the finite list of characteristic polynomials of primitive elements of K that are in ℤK and minimal for the T2 norm; now remove from the list the polynomials whose discriminant do not have minimal absolute value. Note that this condition is restricted to the original list of polynomials with minimal T2 norm and does not imply that the defining polynomial for the field with smallest discriminant belongs to the list !
>
> * To a polynomial P(x) = xn +...+ an ∈ ℝ[x] we attach the sequence S(P) given by |a1|, a1,..., |an|, an. Order the polynomials P by the lexicographic order on the coefficient vectors S(P). Then the output of [polredabs](https://pari.math.u-bordeaux.fr/dochtml/html-stable/General_number_fields.html#se:polredabs) is the smallest polynomial in the above list for that order. In other words, the monic polynomial which is lexicographically smallest with respect to the absolute values of coefficients, favouring negative coefficients to break ties, i.e. choosing x3-2 rather than x3+2.

roed314 commented on src/sage/rings/number_field/number_field.py:
```diff
@@ -2639,6 +2639,102 @@ def absolute_field(self, names):
+        EXAMPLES::
+
+            sage: f = ZZ['x']('x^3 - 12*x^2 + 2*x - 3')
+            sage: NumberField(f, 'a').polred(polredabs=True, isomorphism_map=True)
```
Did you add an underscore to the method name?  If so, this test needs to get updated.

edgarcosta committed:
fix test by adding underscore

edgarcosta committed:
feat(rings/number_field): add polredbest and polredabs methods

Add public polredbest() and polredabs() methods on absolute number
fields, wrapping PARI's polredbest and polredabs respectively. Each
returns an isomorphic absolute number field with a reduced defining
polynomial; passing isomorphism_map=True additionally returns the
isomorphism from self to the new field.

optimized_representation gains an algorithm='polredbest'/'polredabs'
argument (default 'polredbest' preserves prior behavior) so the
existing entry point also exposes the canonical polredabs reduction.

The methods are placed on NumberField_absolute: PARI's polred*
operates on polynomials over Q, so on a relative number field
.absolute_field(...) must be called first.

edgarcosta commented:
Rebased on current `develop` and squashed to a single commit. Changes since the last review:

- `optimized_representation` gains `algorithm='polredbest'|'polredabs'` (default preserves prior behavior).
- Moved `_polred`/`polredbest`/`polredabs` from `NumberField_generic` to `NumberField_absolute`. PARI's polred operates on polynomials over Q, so on a relative number field the previous placement crashed with `PariError`; `.absolute_field(...).polredabs()` is the supported path.
- `_polred` takes `algorithm=...` instead of `polredabs=bool` for consistency.
- Docstring polish: imperative mood, typo fixes, broken `polredbest()` example fixed, `.. SEEALSO::` cross-links.
- Added a TESTS block verifying the returned isomorphism is a valid ring homomorphism for both algorithms.

@user202729 Let me know what you think

edgarcosta committed:
refactor(rings/number_field): rename polred isomorphism_map parameter to map

The boolean flag on _polred, polredbest, and polredabs that controls
whether the change-of-variables isomorphism is also returned was named
isomorphism_map, which is non-standard. Rename it to map, matching the
sibling optimized_representation (which uses a boolean map flag for the
identical polred computation) and the common Sage convention. certificate
is reserved for predicates such as is_isomorphic.

edgarcosta commented:
@mantepse , agreed `isomorphism_map` was non-standard. Renamed it to `map` on `_polred`,
`polredbest`, and `polredabs`. These aren't predicates, so the second return is a
real value (the `self -> K` isomorphism), not a certificate; `map=` matches the
sibling `optimized_representation` and is the common Sage convention. I'd keep
`certificate` for predicates like `is_isomorphic` (separate PR, needs a
deprecation) and leave `both_maps` alone (it means one-vs-both directions). Happy
to open a follow-up issue for the broader unification.
